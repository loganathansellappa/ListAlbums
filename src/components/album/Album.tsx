import React, { useMemo } from 'react';
import { Entry } from '../../@types/Album';
import './Album.scss';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart } from '@fortawesome/free-solid-svg-icons';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../store';
import { add, remove } from '../../states/Albums';

interface AlbumProps {
  album: Entry;
}

export const Album = ({ album }: AlbumProps) => {
  const dispatch = useDispatch();
  const isFavourite = useSelector((state: RootState) =>
    state.favoriteAlbums.favourites?.find(
      (p) => p.id.attributes['im:id'] === album.id.attributes['im:id']
    )
  );

  const setFavourite = () => {
    if (isFavourite) {
      dispatch(remove(album.id.attributes['im:id']));
      return;
    }
    dispatch(add(album));
  };

  const bgImage = useMemo(() => {
    const url = album['im:image']?.at(-1)?.label;
    return url ? <img src={url} alt={album['im:name'].label} /> : null;
  }, [album]);

  return (
    <div className="card">
      {bgImage}
      <div className={'name'}>
        <h3 className={'searchResult'}>{album['im:name'].label}</h3>
      </div>
      <div className="propContainer">
        <div className="props">
          <div className="value">{album.title.label || 'N/A'}</div>
        </div>
        <div className="props">
          <div className="label">Price</div>
          <div className="value">{album['im:price'].label || 'N/A'}</div>
        </div>
      </div>
      <div className="favourite" onClick={setFavourite}>
        <div className={isFavourite ? 'favouriteAlbum' : ''}>
          {!isFavourite ? 'Mark' : ''} Favourite
        </div>
        <div>
          <FontAwesomeIcon
            className={'favouriteIcon'}
            icon={faHeart}
            color={isFavourite ? 'red' : 'white'}
          />
        </div>
      </div>
    </div>
  );
};
