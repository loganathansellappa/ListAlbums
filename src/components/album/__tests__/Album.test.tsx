import {
  mockResource,
  renderWithRouter,
  testStore,
} from '../../..//utils/tests';
import { Album } from '../Album';
import { Provider } from 'react-redux';

describe('Album', () => {
  let element: { container: HTMLElement };
  beforeEach(() => {
    element = renderWithRouter(
      <Provider store={testStore()}>
        <Album album={mockResource().feed.entry[0]} />
      </Provider>
    );
  });

  test('renders component', () => {
    expect(element.container.firstChild).toMatchSnapshot();
    expect(element.container.textContent).toContain('Life After Death');
  });
});
