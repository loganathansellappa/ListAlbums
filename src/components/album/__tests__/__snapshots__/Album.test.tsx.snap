// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Album renders component 1`] = `
<div
  class="card"
>
  <img
    alt="Life After Death"
    src="https://is2-ssl.mzstatic.com/image/thumb/Music112/v4/5d/1b/af/5d1baf4a-496c-be36-b5c1-ba5bcea1c2c0/22UMGIM57228.rgb.jpg/170x170bb.png"
  />
  <div
    class="name"
  >
    <h3
      class="searchResult"
    >
      Life After Death
    </h3>
  </div>
  <div
    class="propContainer"
  >
    <div
      class="props"
    >
      <div
        class="value"
      >
        Life After Death - TobyMac
      </div>
    </div>
    <div
      class="props"
    >
      <div
        class="label"
      >
        Price
      </div>
      <div
        class="value"
      >
        $10.99
      </div>
    </div>
  </div>
  <div
    class="favourite"
  >
    <div
      class=""
    >
      Mark
       Favourite
    </div>
    <div>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-heart favouriteIcon"
        color="white"
        data-icon="heart"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 190.9V185.1C0 115.2 50.52 55.58 119.4 44.1C164.1 36.51 211.4 51.37 244 84.02L256 96L267.1 84.02C300.6 51.37 347 36.51 392.6 44.1C461.5 55.58 512 115.2 512 185.1V190.9C512 232.4 494.8 272.1 464.4 300.4L283.7 469.1C276.2 476.1 266.3 480 256 480C245.7 480 235.8 476.1 228.3 469.1L47.59 300.4C17.23 272.1 .0003 232.4 .0003 190.9L0 190.9z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
